server/src/main/java/org/elasticsearch/index/analysis/CustomAnalyzer.java

```diff
@@ -22,73 +22,65 @@
 import org.apache.lucene.analysis.Analyzer;
 import org.apache.lucene.analysis.TokenStream;
 import org.apache.lucene.analysis.Tokenizer;
+import org.elasticsearch.index.analysis.CustomAnalyzerProvider.AnalyzerComponents;
 
 import java.io.Reader;
 
-public final class CustomAnalyzer extends Analyzer {
+public class CustomAnalyzer extends Analyzer {
 
-    private final String tokenizerName;
-    private final TokenizerFactory tokenizerFactory;
-
-    private final CharFilterFactory[] charFilters;
-
-    private final TokenFilterFactory[] tokenFilters;
-
-    private final int positionIncrementGap;
-    private final int offsetGap;
     private final AnalysisMode analysisMode;
+    protected volatile AnalyzerComponents components;
 
     public CustomAnalyzer(String tokenizerName, TokenizerFactory tokenizerFactory, CharFilterFactory[] charFilters,
-            TokenFilterFactory[] tokenFilters) {
-        this(tokenizerName, tokenizerFactory, charFilters, tokenFilters, 0, -1);
+                          TokenFilterFactory[] tokenFilters) {
+        this(new AnalyzerComponents(tokenizerName, tokenizerFactory, charFilters, tokenFilters, 0, -1), GLOBAL_REUSE_STRATEGY);
     }
 
-    public CustomAnalyzer(String tokenizerName, TokenizerFactory tokenizerFactory, CharFilterFactory[] charFilters,
-            TokenFilterFactory[] tokenFilters, int positionIncrementGap, int offsetGap) {
-        this.tokenizerName = tokenizerName;
-        this.tokenizerFactory = tokenizerFactory;
-        this.charFilters = charFilters;
-        this.tokenFilters = tokenFilters;
-        this.positionIncrementGap = positionIncrementGap;
-        this.offsetGap = offsetGap;
-        // merge and transfer token filter analysis modes with analyzer
-        AnalysisMode mode = AnalysisMode.ALL;
-        for (TokenFilterFactory f : tokenFilters) {
-            mode = mode.merge(f.getAnalysisMode());
-        }
-        this.analysisMode = mode;
+    CustomAnalyzer(AnalyzerComponents components, ReuseStrategy reuseStrategy) {
+        super(reuseStrategy);
+        this.components = components;
+        this.analysisMode = calculateAnalysisMode(components);
     }
 
+    /**
+     * TODO: We should not expose functions that return objects from the <code>current</code>,
+     * only the full {@link AnalyzerComponents} should be returned
+     */
```

Comment:
@jimczi I left this comment of yours because I'm not sure what would change if we returned only the full AnalyzerComponents. Clients that now usse the following getters would still need to access indovidual components on it with the potential of changing their state, or am I missing something? Also I think it would be good to keep the API of CustomAnalyzer stable for now since it might be one of the more popular APIs for use in custom plugins etc..?

Reply:
We change the entire AnalyzerComponents on update so the idea exposed here is that users should not access the underlying object individually. Instead they should retrieve the AnalyzerComponents once (it is a volatile object) and work with the underlying objects since they are final:
````
  final AnalyzerComponents components = analyzer.getComponents();
  // we need charFilters and tokenFilters to do something so we need to ensure that they come 
  // from the same component
  mixFilters(components.charFilters, components.tokenFilters);
````